pkg/apis/serving/v1/serviceorchestrator_types.go

```diff
@@ -0,0 +1,176 @@
+var (
+	// Check that we can create OwnerReferences to a ServiceOrchestrator.
+	_ kmeta.OwnerRefable = (*ServiceOrchestrator)(nil)
+
+	// Check that the type conforms to the duck Knative Resource shape.
+	_ duckv1.KRShaped = (*ServiceOrchestrator)(nil)
+)
+
+// TargetRevision holds the information of the revision for the current stage.
+type TargetRevision struct {
+	// RevisionName indicates RevisionName.
+	// +optional
+	RevisionName string `json:"revisionName,omitempty"`
+
+	// Direction indicates up or down.
+	// +optional
+	Direction string `json:"direction,omitempty"`
+
+	// TargetReplicas indicates an estimated number of replicas.
+	// +optional
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// IsLatestRevision indicates whether it is the last revision or not.
+	// +optional
+	IsLatestRevision *bool `json:"IslatestRevision,omitempty"`
+
+	// Percent indicates that percentage based routing should be used and
+	// the value indicates the percent of traffic that is routed to this
+	// Revision. `0` (zero) mean no traffic, `100` means all
+	// traffic.
+	// When percentage based routing is being used the follow rules apply:
+	// - the sum of all percent values must equal 100
+	// - when not specified, the implied value for `percent` is zero for
+	//   that particular Revision
+	// +optional
+	Percent *int64 `json:"percent,omitempty"`
+
+	// MinScale sets the lower bound for the number of the replicas.
+	// +optional
+	MinScale *int32 `json:"minScale,omitempty"`
+
+	// MaxScale sets the upper bound for the number of the replicas.
+	// +optional
+	MaxScale *int32 `json:"maxScale,omitempty"`
+}
+
+// StageTarget holds the information of all revisions during the transition for the current stage.
+type StageTarget struct {
+	// StageTargetRevisions holds the configured traffic distribution for the current stage.
+	// These entries will always contain RevisionName references.
+	// +optional
+	StageTargetRevisions []TargetRevision `json:"stageTargetRevisions,omitempty"`
+
+	// TargetFinishTime indicates target time to complete this target.
+	// +optional
+	TargetFinishTime apis.VolatileTime `json:"targetFinishTime,omitempty"`
+}
+
+// ServiceOrchestratorSpec holds the desired state of the ServiceOrchestrator (from the client).
+type ServiceOrchestratorSpec struct {
+	StageTarget `json:",inline"`
+
+	// StageTarget holds the information of all revisions during the transition for the current stage.
+
+	// TargetRevisions holds the information of the target revisions in the final stage.
+	// These entries will always contain RevisionName references.
+	// +optional
+	TargetRevisions []TargetRevision `json:"targetRevisions,omitempty"`
+
+	// InitialRevisions holds the information of the initial revisions in the initial stage.
+	// These entries will always contain RevisionName references.
+	// +optional
+	InitialRevisions []TargetRevision `json:"initialRevisions,omitempty"`
```

Comment:
Can you help document an example how `InitialRevision` and `TargetRevision` are used?

Reply:
Yes.
1. When there is no revision available, the InitialRevisions will be empty, and the TargetRevisions will be empty. 
2. When there is an existing revision, and we create a new revision, InitialRevisions will be set to the existing revision with 100% traffic, and TargetRevisions will be set to the new revision with 100% traffic.
3. During the transition from the old to new revision, TargetRevisions and InitialRevisions remain the same as TargetRevisions pointing to the new revision of the ultimate goal, and InitialRevision pointing to the old revision of the initial status.
4.  When the transition is over, InitialRevisions and TargetRevisions will be reset to empty.

Reply:
ok,  good to leave this as comments on the defined fields.